### PR TITLE
Added a Statsd module

### PR DESCRIPTION
**Pull request by simong**
Currently tracking:
 - HTTP:
    * Counting per route and total
    * Timings per route

The idea is that we'll use graphite for system specific information such as (global) API response times, DB counts & rw ratios, server disk usage, etc..

Graphite's storage backend doesn't really allow us to collect metrics on a per-app basis. (There's a sparse feature which sounds useful, but doesn't save us much (~1.1GB out of 3.5GB). We should use a more generic BI tool that allows us to do this segmentation such as Mixpanel, Google analytics, Tableau, etc..

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -102,3 +100,18 @@ config.servers = {
 config.signing = {
     'key': 'The default signing key, please change me.'
 };
+
+/*!
+ * The statsd related configuration
+ *
+ * @property  {Boolean}     enabled                 Whether the application should push telemetry data into statsd
+ * @property  {String}      host                    The hostname where statsd can be reached
```
Comment (nicolaasmatthijs) [inline config.js]: `on which`

File: config.js [hunk 2]
```diff
@@ -102,3 +100,18 @@ config.servers = {
...
+ * @property  {Number}      port                    The port where statsd can be reached
```
Comment (nicolaasmatthijs) [inline config.js]: `on which`

File: node_modules/gh-core/lib/server.js [hunk 3]
```diff
@@ -56,6 +57,40 @@ var setUpServer = module.exports.setUpServer = function(port, config) {
...
+        var start = Date.now();
+        res.on('finish', function() {
+            var url = req.url;
+            var code = res.statusCode;
+
+            // Try to get a more generalised URL. We try to re-use the
+            // Expressjs routing logic and convert `/api/user/1234` into
+            // something that can be stored in Statsd: `api_user_id`
+            if (req.route && req.route.path) {
+                url = req.route.path;
```
Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/server.js]: Don't we have a `telemetryUrl` on the request object that we can use for this?

File: node_modules/gh-core/lib/server.js [hunk 4]
```diff
@@ -56,6 +57,40 @@ var setUpServer = module.exports.setUpServer = function(port, config) {
...
+                // Remove the first and last slash
+                url = url.replace(/^\/|\/$/g, '');
+
+                // Replace any remaining slashes or dots with underscores
+                url = url.replace(/\/|\./g, '_');
+            }
+
+            // Track the time and response code for the route
+            var requestName = 'http.' + req.method + '.' + url;
+            StatsdAPI.getInstance().increment(requestName + '.' + code + '.count');
```
Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/server.js]: Can we do the `getInstance()` above so we don't have to repeat it?

File: node_modules/gh-statsd/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# The Grasshopper Event Engine statsd module
+
+The statsd module is responsible for interacting with Statsd
```
Comment (nicolaasmatthijs) [inline node_modules/gh-statsd/README.md]: Captain Obvious

File: node_modules/gh-statsd/lib/api.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2014 "Fronteer LTD"
```
Comment (nicolaasmatthijs) [inline node_modules/gh-statsd/lib/api.js]: We should update these all to 2015 :|

File: node_modules/gh-statsd/README.md (hunk 5 above)
Comment (simong) [inline node_modules/gh-statsd/README.md]: This is your captain speaking

File: node_modules/gh-statsd/lib/api.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+var init = module.exports.init = function(config) {
+    if (config.enabled) {
+        metrics = new lynx(config.host, config.port, {'scope': config.scope});
+    }
+};
+
+/**
+ * Get a statsd object. If statsd is disabled, a stubbed object will be returned
+ *
+ * @return {Lynx} An object that allows you to push data to statsd
```
Comment (nicolaasmatthijs) [inline node_modules/gh-statsd/lib/api.js]: Need some spacing after `}`

**Comment by nicolaasmatthijs** [comment]
Looks good. Left some minor feedback. Back to @simong for follow-up

**Comment by simong** [comment]
Followed-up, back to @nicolaasmatthijs.

The years can be done separately

**Comment by simong** [comment]
I'll try to attach a grafana dashboard as well

**Comment by simong** [comment]
Added an example dashboard:

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @simong for follow-up

**Comment by nicolaasmatthijs** [comment]
@simong : Looks like there is a merge conflict here

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @simong for follow-up

**Comment by nicolaasmatthijs** [comment]
Looks good. Merged.

**Pull request merged**